### PR TITLE
Report failed design time builds

### LINKED ISSUE #8318: A failed design-time build is not reported to the user and can cause errors during build that are confusing
## Visual Studio Version
Any

## Summary
A user can add a custom target to do something during their build.  However, most users do not realize they need to exclude their target during design-time builds.  A failing design-time build gives NuGet an incomplete nomination and so restore will succeed even though the design-time build failed.  There's nothing bubbled up to the user as to what happened so generally users blame NuGet for not restoring.

## Steps to Reproduce

1.  Add a custom target to a project:
  ```xml
    <Target Name="BreakDesignTimeBuild" DependsOnTargets="TargetThatDoesNotExistExceptAfterRestoreSucceeds">
    </Target>
  ```
2.  Load project in Visual Studio for the first time

## Expected Behavior
Visual Studio/Project System should notify the user that the design-time build failed and possibly tell NuGet about it in the nomination details.  Or perhaps nomination/restore should not take place until the design-time build targets succeed.

## Actual Behavior
Projects load, restore succeeds, but build errors occur. 

## User Impact
Any user who adds custom targets that break design time build is affected.  At least two customers have reported this issue in the last three months.

## Workaround
Users who add custom targets should condition them so they don't affect design-time builds:
```xml
Condition="'$(DesignTimeBuild)' != 'true'"
```

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #8318

Our design-time build logger observes builds and reports on the result, and the most expensive targets that ran.

Today, when a DTB fails, the user is not notified directly. An error will be present in the Error List, but it's often not clear that the error causes various features of the IDE to fail.

With this change, when a DTB failure occurs the user is shown an info bar explaining:

> An error occurred during a design-time build. Some features may not work correctly. See the Error List for details.

In addition, the error count, and names of targets that throw errors, are reported for investigation.

Previously all target names were hashed. With this change, we won't hash targets from files shipped in the VS install directory, or in the `dotnet/sdk` folder of `Program Files`, making investigations of performance problems in Microsoft-provided MSBuild targets easier.

Hashed target names are stored in a lookup, to prevent repeated hash calculation during DTBs.

Some code formatting and updates.

---

We also use shorter hashed values in telemetry . Previously, target name hashes were converted to strings via `BitConverter.ToString(byte[])`, which would produce a string like:

> E3-B0-C4-42-98-FC-1C-14-9A-FB-F4-C8-99-6F-B9-24-27-AE-41-E4-64-9B-93-4C-A4-95-99-1B-78-52-B8-55

This is a very long string, and takes up more space in the telemetry system than is needed. With this change, the above string would instead be:

> e3b0c44298fc1c14

This holds 64 bits rather than 128, and the removal of dashes removes redundant characters.

Because of the changes to telemetry data, we bump the version number of the property. This will allow any existing queries to continue working (I'm not aware of any) and make it easier to write queries using the new format in future.

---

For example, given a project:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net8.0</TargetFramework>
    <CollectUpToDateCheckBuiltDesignTimeDependsOn>Crash;TargetThatDoesNotExist</CollectUpToDateCheckBuiltDesignTimeDependsOn>
  </PropertyGroup>

  <Target Name="Crash">
    <Error Text="Oops, I did it again!" ContinueOnError="true" Code="OOPS001" />
  </Target>

</Project>
```

An info bar will be displayed:

image

And the error list displays:

image

Indeed, as the info bar message states, several IDE features don't work correctly. For example, the Roslyn language service was not initialized due to the failure, so the editor doesn't provide highlighting, RichNav, IntelliSense, etc.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/project-system/pull/9583)

### CODE REVIEW

**Comment by jeffkl** [comment]
@drewnoakes I'm so excited this is being worked on, thank you!  Just surfacing the error is a huge step forward.  Do you think it would be possible to alert users to the fact that they may not want to be running their target during a design-time build?  Something to the effect 

> If you do not want this target to run during design time builds, you can set a condition on the target.  See https://github.com/dotnet/project-system/blob/main/docs/design-time-builds.md for more info

The core targets fail sometimes, but in my experience its people's own custom targets that are just running during design time builds when they didn't really intend them too.  I don't see how we could possibly differentiate between a target that is supposed to run and is failing vs a target that probably shouldn't be running at all though.  I just think if we can give users any hint at all that it would be extremely helpful.  What do you think?

**Review by JakeRadMSFT — APPROVED** [review]

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DesignTimeBuilds/DesignTimeBuildLoggerProvider.cs [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        return Task.FromResult(loggers);
+    }
+
+    private sealed class DesignTimeTelemetryLogger(ITelemetryService telemetryService, IInfoBarService infoBarService, IProjectFaultHandlerService projectFaultHandlerService, ConfiguredProject project) : ILogger
+    {
+        /// <summary>
+        /// Stores hashed target names so that we don't recompute these hashes for every project configuration.
+        /// The set of target names is relatively small.
+        /// </summary>
+        private static ImmutableDictionary<(string Target, bool IsMicrosoft), string> s_hashByTargetName = ImmutableDictionary<(string Target, bool IsMicrosoft), string>.Empty;
```
Comment (melytc) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DesignTimeBuilds/DesignTimeBuildLoggerProvider.cs:44]: nit: is the bool `IsMicrosoft` refer to internal users? if so, can we rename it to `IsInternalUser` or something like that?

Comment (melytc) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DesignTimeBuilds/DesignTimeBuildLoggerProvider.cs:44]: or does it refer to a Microsoft's target?

**Review by melytc — APPROVED** [review]
Looks great!

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DesignTimeBuilds/DesignTimeBuildLoggerProvider.cs (hunk 1 above)
Comment (melytc) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DesignTimeBuilds/DesignTimeBuildLoggerProvider.cs:44]: After reading all your changes, I think it makes sense now :)

**Comment by drewnoakes** [comment]
> Do you think it would be possible to alert users to the fact that they may not want to be running their target during a design-time build?

@jeffkl I'm going to monitor this for occurrences. Running user targets during DTB is one scenario, but there are others too.

Once we have some more data about this, I suspect we'll end up writing a doc on this and have a link to the docs from the info bar.

Will be keeping an eye out for feedback about this, and on the telemetry to see how widespread this is.

**Comment by jeffkl** [comment]
Sounds good, thanks again for this improvement!!

**Pull request merged**